### ISSUE TITLE
lint impl on trait object

### ISSUE THREAD

**Issue by durka** [Manishearth/rust-clippy#1612]
Example code:

```rust
trait Foo {}

impl Foo {
    fn bar(&self) {
        /* whatever */
    }
}
```

The lint would be "warning: this adds a method to a trait object type, which is rarely intended. Consider a blanket impl of `Foo` instead."

Advanced mode: rewrite the impl header to suggest `impl<T> Foo for T`.

**Comment by sgrif** [comment]
Another semi-related one I've seen when working with an API that passes trait objects around:

```
trait FooExtensions {
    fn bar(&self);
}

impl<'a> FooExtensions for Foo + 'a {
    fn bar(&self) {}
}
```

It'd be nice to have a lint that recommends changing that to `impl<T: Foo + ?Sized> FooExtensions for T`
